### PR TITLE
Rewrite style images with a good dose of generics 💉

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16859)

### CODE REVIEW

**Comment by nox** [comment]
Try run with expectation changes included: https://treeherder.mozilla.org/#/jobs?repo=try&revision=356cac446ec23bdc85b7593af71f43cb90c655ae

File: components/style/values/generics/image.rs [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+                        dest.write_str(", ")?;
+                        shape.to_css(dest)?;
+                    }
+                }
+                false
+            },
+        };
+        for item in &self.items {
+            if !skip_comma {
+                try!(dest.write_str(", "));
```
Comment (emilio) [inline components/style/values/generics/image.rs]: nit: The rest of the function uses `?`, perhaps this should too.

File: components/style/values/generics/image.rs [hunk 2]
```diff
@@ -0,0 +1,637 @@
...
+            },
+        }
+    }
+}
+
+impl<G, R> HasViewportPercentage for Image<G, R>
+    where G: HasViewportPercentage
+{
+    fn has_viewport_percentage(&self) -> bool {
+        if let Image::Gradient(ref gradient) = *self {
```
Comment (emilio) [inline components/style/values/generics/image.rs]: nit: I think `match` is nicer than `if let { ... } else { }`, but feel free to disagree on this :)

**Review by emilio — APPROVED** [review]
Looks quite fancy! r=me with nits

**Pull request merged**